stdlib/src/collections/counter.mojo

```diff
@@ -76,3 +73,363 @@ struct Counter[V: KeyElement](
             count: The count to store in the Counter.
         """
         self._data[value] = count
+
+    fn __iter__(ref [_]self: Self) -> _DictKeyIter[V, Int, __lifetime_of(self)]:
+        """Iterate over the keyword dict's keys as immutable references.
+
+        Returns:
+            An iterator of immutable references to the Counter values.
+        """
+        return self._data.__iter__()
+
+    fn __len__(self) -> Int:
+        """The number of elements currently stored in the Counter."""
+        return len(self._data)
+
+    fn __bool__(self) -> Bool:
+        """Check if the Counter is empty or not.
+
+        Returns:
+            `False` if the Counter is empty, `True` if there is at least one element.
+        """
+        return len(self).__bool__()
+
+    fn __eq__(self, other: Self) -> Bool:
+        """Check if all counts agree. Missing counts are treated as zero.
+
+        Args:
+            other: The other Counter to compare to.
+
+        Returns:
+            True if the two Counters are equal, False otherwise.
+        """
+
+        @parameter
+        @always_inline
+        fn is_eq(keys: _DictKeyIter[V, Int, _]) -> Bool:
+            for e_ref in keys:
+                var e = e_ref[]
+                if self.get(e, 0) != other.get(e, 0):
+                    return False
+            return True
+
+        return is_eq(self.keys()) and is_eq(other.keys())
+
+    fn __ne__(self, other: Self) -> Bool:
+        """Check if all counts disagree. Missing counts are treated as zero.
+
+        Args:
+            other: The other Counter to compare to.
+
+        Returns:
+            True if the two Counters are not equal, False otherwise.
+        """
+        return not self == other
+
+    fn __le__(self, other: Self) -> Bool:
+        """Check if all counts are less than or equal to the other Counter.
+
+        Args:
+            other: The other Counter to compare to.
+
+        Returns:
+            True if all counts are less than or equal to the other Counter, False otherwise.
+        """
+
+        @parameter
+        @always_inline
+        fn is_le(keys: _DictKeyIter[V, Int, _]) -> Bool:
+            for e_ref in keys:
+                var e = e_ref[]
+                if self.get(e, 0) > other.get(e, 0):
+                    return False
+            return True
+
+        return is_le(self.keys())
+
+    fn __lt__(self, other: Self) -> Bool:
+        """Check if all counts are less than in the other Counter.
+
+        Args:
+            other: The other Counter to compare to.
+
+        Returns:
+            True if all counts are less than in the other Counter, False otherwise.
+        """
+        return self <= other and self != other
```

Comment:
Should we rely on loop fusion here?

Reply:
Not sure if it's worth it. I think currently is much cleaner to read. Not sure if @laszlokindrat has some opinion